### PR TITLE
Implement HTMLCanvasElement.probablySupportsContext

### LINKED ISSUE #11403: Implement HTMLCanvasElement.probablySupportsContext
Blocks #2968.

This just checks if `HTMLCanvasElement.GetContext` would return `None`, if so `probablySupportsContext` should return false, otherwise it will return true.

---
Spec: https://html.spec.whatwg.org/multipage/scripting.html#dom-canvas-probablysupportscontext
Code: `components/script/dom/webidl/HTMLCanvasElement.webidl` `components/script/dom/webidl/htmlcanvaselement.rs`

### PR DESCRIPTION
**Pull request by mskrzypkows**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11403

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11461)

### CODE REVIEW

**Comment by nox** [comment]
r? @emilio 

I think this change is incorrect, because `GetContext` will cache the context, so if you try to check whether you can create a `2d` context, and then try to get for real a `canvas` context, you won't get it.

File: components/script/dom/htmlcanvaselement.rs [hunk 1]
```diff
@@ -249,6 +249,18 @@ impl HTMLCanvasElementMethods for HTMLCanvasElement {
         }
     }
 
+    // https://html.spec.whatwg.org/multipage/#dom-canvas-probablysupportscontext
+    fn ProbablySupportsContext(&self,
+                  cx: *mut JSContext,
+                  id: DOMString,
+                  attributes: Vec<HandleValue>) -> bool {
+        if self.GetContext(cx, id, attributes).is_some() {
```
Comment (dzbarsky) [inline components/script/dom/htmlcanvaselement.rs]: just `return self.GetContext(cx, id, attributes).is_some()`

**Comment by emilio** [comment]
@nox is right here, we can't do this that way.

In fact, there's no real support for this method in any other browser (and it's arguable), see [this](https://wiki.mozilla.org/Platform/GFX/CanvasThingsWeDontPlanToSupport) for example.

In any case, if we plan to implement this, we should use an heuristic, and not calling `GetContext()` (since the whole point of this API is avoiding the `GetContext()` overhead).

In my opinion, we should return true if the `id` is one of those we support (i.e. `webgl`, `experimental-webgl` or `2d`), and *maybe* we want to make it return false if getting the `WebGLContextAttributes` we see that `antialias` is true (it's something we don't support right now), but that's an extra thing to change when we support it, so... I'd just check the `id`.

That being said, I'm a bit skeptical about the utility of this method.

-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> r? @emilio 
> 
> 
> 
> I think this change is incorrect, because `GetContext` will cache the context, so if you try to check whether you can create a `2d` context, and then try to get for real a `canvas` context, you won't get it.
</details>


---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11461#-:-KIr1nZIXFZV2eomQM0T:-407728463)*

**Comment by Ms2ger** [comment]
We should probably figure out why no other browser implemented this; it's not particularly likely to be implementation difficulty.

**Comment by mskrzypkows** [comment]
Here is whatwg thread: https://lists.w3.org/Archives/Public/public-whatwg-archive/2013Jun/0098.html

**Comment by mskrzypkows** [comment]
Is there any decision about this task?

**Comment by emilio** [comment]
I just filled https://github.com/whatwg/html/pull/1459

**Comment by emilio** [comment]
The spec has been modified in order to no longer include this method.

Thanks for the contribution though @mskrzypkows, it's been crucial to raise the issue! :-)

**Pull request closed without merging**